src/hilok.hpp

```diff
@@ -1,33 +1,260 @@
+#include <memory>
 #include <string>
 #include <string_view>
+#include <shared_mutex>
+#include <unordered_map>
+#include <map>
+#include <vector>
+#include <thread>
+#include <atomic>
+
+class HiErr : public std::runtime_error {
+    using std::runtime_error::runtime_error;
+};
+
+
+class HiMutex {
+    
+public:
+    std::thread::id m_ex_id;
+    int m_ex_cnt;
+    std::map<std::thread::id, int> m_shared;
+    std::shared_timed_mutex m_mut;
+    std::mutex m_internal_mut;
+    bool m_recursive;
+
+    HiMutex(bool recursive) : m_ex_cnt(0), m_recursive(recursive) {
+    }
+
+    bool _has_lock() {
+        return m_ex_cnt > 0 && m_ex_id == std::this_thread::get_id();
+    }
+
+    bool _do_lock() {
+        if (!m_recursive)
+            return false;
+        std::lock_guard<std::mutex> guard(m_internal_mut);
+        if (_has_lock()) {
+            ++m_ex_cnt;
+            return true;
+        }
+        return false;
+    }
+
+    bool internal_shared_lock(bool block, double secs) {
+        bool ret;
+        if (!block) {
+            ret = m_mut.try_lock_shared();
+        } else if (secs != 0.0) {
+            ret = m_mut.try_lock_shared_for(std::chrono::duration<double>(secs));
+        } else {
+            m_mut.lock_shared();
+            ret = true;
+        }
+        return ret;
+    }
+ 
+    bool unsafe_clone_lock_shared(HiMutex &src, bool block, double secs) {
+        // lock number of times matching src
+        for (auto it = src.m_shared.begin(); it!= src.m_shared.end(); ++ it) {
```

Comment:
nit: think you should be able to do

```suggestion
        for (const auto &[tid, cnt] : src.m_shared) {
```